### ISSUE TITLE
Missing XMLNode addOption

### ISSUE THREAD

**Issue by adler-j** [astra-toolbox/astra-toolbox#44]
When installing astra and having everything set up, I get the following error when running `import astra`

```
ImportError: astra/utils.so: undefined symbol: _ZN5astra7XMLNode9addOptionENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_
```

demangling this we get

```
astra::XMLNode::addOption(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
```

or more simply

```
astra::XMLNode::addOption(std::string, std::string)
```

EDIT:

checking `utils.so` with nm we find:

```
U _ZN5astra7XMLNode9addOptionENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_
```

so it is not defined in that file, in `libastra.so` a similar function is defined

```
T _ZN5astra7XMLNode9addOptionESsSs
```
```
astra::XMLNode::addOption(std::string, std::string)
```

but this is not the same. It would seem that parts of the library is built with c++11 support, and some parts are not.

**Comment by dmpelt** [comment]
Interesting... How did you compile/install ASTRA and what OS and GCC version are you running? 

The `std::__cxx11` things in the error message could be related to the dual C++ ABI of recent GCC versions (see [here](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html)). Maybe the ASTRA c++ library is compiled with the old ABI, but the python compilation defaults to the new ABI? I think this could result in the errors shown. The "Troubleshooting" section in the link above also suggests this. In this case, a solution would be to recompile everything with the same settings (either both with the old ABI, or both with the new ABI).

**Comment by adler-j** [comment]
OS is Ubuntu 15.10 (I'm not admin of this machine so not 100% about everything, just helping out). Anyway GCC 5.2.1 is default, but since this has issues with CUDA we opt for 4.8 by using the `CC` and `CXX` flags.

I have recompiled all of astra (like 10 times), but it seems to persist. I suspect that you perhaps do not properly pass flags to the cython compiler (which creates `utils.so`), perhaps using the system default instead of the one specified by the user?

**Comment by adler-j** [comment]
Update: the above is indeed the issue, by setting

```python
os.environ['CC'] = 'gcc-4.8'
os.environ['CXX'] = 'g++-4.8'
```

in `builder.py`, the issue is fixed. Perhaps you can make sure that `builder.py` is somehow correctly fed the correct flags as given by the user. That should fix this.

**Comment by dmpelt** [comment]
Ok, nice to know that that is the issue!

How did you pass the flags to the builder system? I think things like `CXXFLAGS` should be passed to the Python build script correctly, but I might have to check it.

**Comment by adler-j** [comment]
I tried adding `CXX = $(CXX)` in the makefile, but it made no difference, so it has to go somewhere else.

I solved it by putting the lines 

```
os.environ['CC'] = 'gcc-4.8'
os.environ['CXX'] = 'g++-4.8'
```

into the `builder.py` script at the top.

It seems from the makefile that you already pass `CXXFLAGS` this way.

**Comment by dmpelt** [comment]
Now I don't really understand why the C++ part of astra is building without the new ABI in your setup. Do you specify `CC`/`CXX` during `configure`?

**Comment by adler-j** [comment]
The C++ part is building without the new ABI since GCC 4.8 uses std=c++98 by default while GCC 5.2 uses std=c++11. Since i specify the `CXX`/`CC` parameters as GCC 4.8 to `configure` astra is built with GCC 4.8. However, when make calls `python builder.py build ...`, you dont pass this parameter choice through to cython, causing cython to use the system default (which would be GCC 5.2, thus using the new ABI).

**Comment by dmpelt** [comment]
Thanks for the extra info! 

I looked at the code, and indeed, the `CXX`/`CC` environment variable is not passed to the python compilation step. Although I could not reproduce the specific issue with the new ABI, I could see that the python compilation was using the system default gcc instead of the one supplied by `CXX`. A fix can be found in PR #46. 

Can you try that PR to see if it fixes the ABI issue?

**Comment by adler-j** [comment]
I'll try out your PR later today! Thank you for the fix.

**Comment by adler-j** [comment]
PR works well for me at this computer.

**Comment by adler-j** [comment]
This was solved by #46. Closing.
